examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java

```diff
@@ -0,0 +1,248 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.examples.snippets;
+
+import com.google.api.services.bigquery.model.TableFieldSchema;
+import com.google.api.services.bigquery.model.TableReference;
+import com.google.api.services.bigquery.model.TableRow;
+import com.google.api.services.bigquery.model.TableSchema;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import org.apache.beam.sdk.Pipeline;
+import org.apache.beam.sdk.coders.AvroCoder;
+import org.apache.beam.sdk.coders.DefaultCoder;
+import org.apache.beam.sdk.io.gcp.bigquery.BigQueryIO;
+import org.apache.beam.sdk.io.gcp.bigquery.BigQueryIO.Write.CreateDisposition;
+import org.apache.beam.sdk.io.gcp.bigquery.BigQueryIO.Write.WriteDisposition;
+import org.apache.beam.sdk.transforms.Create;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.MapElements;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.join.CoGbkResult;
+import org.apache.beam.sdk.transforms.join.CoGroupByKey;
+import org.apache.beam.sdk.transforms.join.KeyedPCollectionTuple;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.TupleTag;
+import org.apache.beam.sdk.values.TypeDescriptor;
+import org.apache.beam.sdk.values.TypeDescriptors;
+
+/**
+ * Code snippets used in webdocs.
+ */
+public class Snippets {
+
+  @DefaultCoder(AvroCoder.class)
+  static class Quote {
+    final String source;
+    final String quote;
+
+    public Quote() {
+      this.source = "";
+      this.quote = "";
+    }
+    public Quote(String source, String quote) {
+      this.source = source;
+      this.quote = quote;
+    }
+  }
+
+  /** Using a Read and Write transform to read/write from/to BigQuery. */
+  public static void modelBigQueryIO(Pipeline p) {
+    modelBigQueryIO(p, "", "", "");
+  }
+
+  public static void modelBigQueryIO(Pipeline p,
+      String writeProject, String writeDataset, String writeTable) {
+    {
+      // [START BigQueryTableSpec]
+      String tableSpec = "clouddataflow-readonly:samples.weather_stations";
+      // [END BigQueryTableSpec]
+    }
+
+    {
+      // [START BigQueryTableSpecWithoutProject]
+      String tableSpec = "samples.weather_stations";
+      // [END BigQueryTableSpecWithoutProject]
+    }
+
+    {
+      // [START BigQueryTableSpecObject]
+      TableReference tableSpec = new TableReference()
+          .setProjectId("clouddataflow-readonly")
+          .setDatasetId("samples")
+          .setTableId("weather_stations");
+      // [END BigQueryTableSpecObject]
+    }
+
+    {
+      String tableSpec = "clouddataflow-readonly:samples.weather_stations";
+      // [START BigQueryReadTable]
+      PCollection<Double> maxTemperatures = p
+          .apply(BigQueryIO.readTableRows().from(tableSpec))
```

Comment:
Suggestion: `readTableRows()` has poor performance - consider including also a snippet of `read(SerializableFunction)` which performs much better. It's also definitely a better fit for what you're doing here (immediately extracting some information from the row). In fact I'd recommend it for *all* snippets reading from BigQuery, and maybe include just 1 snippet with readTableRows.

Reply:
Got it, changing